Developer/Default.xctoolchain/cctools/ar/ar.1
.\"	$OpenBSD: ar.1,v 1.2 1996/06/26 05:31:17 deraadt Exp $
.\"	$NetBSD: ar.1,v 1.7 1995/08/18 15:05:11 pk Exp $
.\"
.\" Copyright (c) 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Hugh Smith at The University of Guelph.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)ar.1	8.1 (Berkeley) 6/29/93
.\"
.Dd July 27, 2005
.Dt AR 1 
.Os Darwin
.Sh NAME
.Nm ar 
.Nd create and maintain library archives
.Sh SYNOPSIS
.Nm ar
.Fl d 
.Op Fl \TLsv 
.Ar archive file ...
.Nm ar
.Fl m
.Op Fl \TLsv 
.Ar archive file ...
.Nm ar
.Fl m
.Op Fl abiTLsv 
.Ar position archive file ...
.Nm ar
.Fl p
.Op Fl \TLsv
.Ar archive
.Op Ar file ...
.Nm ar
.Fl q
.Op Fl cTLsv
.Ar archive file ...
.Nm ar
.Fl r
.Op Fl cuTLsv
.Ar archive file ...
.Nm ar
.Fl r
.Op Fl abciuTLsv
.Ar position archive file ...
.Nm ar
.Fl t
.Op Fl \TLsv
.Ar archive
.Op Ar file ...
.Nm ar
.Fl x
.Op Fl ouTLsv
.Ar archive
.Op Ar file ...
.Sh DESCRIPTION
The
.Nm ar
utility creates and maintains groups of files combined into an archive.
Once an archive has been created, new files can be added and existing
files can be extracted, deleted, or replaced.
.Pp
Files are named in the archive by a single component, i.e., if a file
referenced by a path containing a slash (``/'') is archived it will be
named by the last component of that path.
When matching paths listed on the command line against file names stored
in the archive, only the last component of the path will be compared.
.Pp
All informational and error messages use the path listed on the command
line, if any was specified, otherwise the name in the archive is used.
If multiple files in the archive have the same name, and paths are listed
on the command line to ``select'' archive files for an operation, only the
.Em first
file with a matching name will be selected.
.Pp
The normal use of
.Nm ar
is for the creation and maintenance of libraries suitable for use with
the loader (see
.Xr ld 1 )
although it is not restricted to this purpose.
The options are as follows:
.Bl -tag -width indent
.It Fl a
A positioning modifier used with the options 
.Fl r 
and 
.Fl m .
The files are entered or moved
.Em after
the archive member
.Ar position ,
which must be specified.
.It Fl b
A positioning modifier used with the options
.Fl r 
and 
.Fl m .
The files are entered or moved
.Em before
the archive member
.Ar position ,
which must be specified.
.It Fl c
Whenever an archive is created, an informational message to that effect
is written to standard error.
If the 
.Fl c
option is specified,
.Nm ar
creates the archive silently.
.It Fl d
Delete the specified archive files.
.It Fl i
Identical to the 
.Fl b
option.
.It Fl m
Move the specified archive files within the archive.
If one of the options 
.Fl a , 
.Fl b 
or 
.Fl i
are specified, the files are moved before or after the
.Ar position
file in the archive.
If none of those options are specified, the files are moved
to the end of the archive.
.It Fl o
Set the access and modification times of extracted files to the
modification time of the file when it was entered into the archive.
This will fail if the user is not the owner of the extracted file
or the super-user.
.It Fl p
Write the contents of the specified archive files to the standard output.
If no files are specified, the contents of all the files in the archive
are written in the order they appear in the archive.
.It Fl q
(Quickly) append the specified files to the archive.
If the archive does not exist a new archive file is created.
Much faster than the 
.Fl r
option, when creating a large archive
piece-by-piece, as no checking is done to see if the files already
exist in the archive.
.It Fl r
Replace or add the specified files to the archive.
If the archive does not exist a new archive file is created.
Files that replace existing files do not change the order of the files
within the archive.
New files are appended to the archive unless one of the options 
.Fl a ,
.Fl b
or 
.Fl i
is specified.
.It Fl T
Select and/or name archive members using only the first fifteen characters
of the archive member or command line file name.
The historic archive format had sixteen bytes for the name, but some
historic archiver and loader implementations were unable to handle names
that used the entire space.
This means that file names that are not unique in their first fifteen
characters can subsequently be confused.
A warning message is printed to the standard error output if any file
names are truncated.
(See
.Xr ar 5
for more information.)
.It Fl L
Used the extended format to allow long archive member names.  This is the
default.
.It Fl s
Write an object-file index into the archive, or update an existing one, even if
no other change is made to the archive.  You may use this modifier flag either
with any operation, or alone.  Running `ar s' on an archive is equivalent to
running `ranlib' on it.
.It Fl S
Do not generate an archive symbol table.  This can speed up building a large
library in several steps.  The resulting archive can not be used with the
linker.  In order to build a symbol table, you must omit the S modifier on the
last execution of ar, or you must run ranlib on the archive.
.It Fl t
List the specified files in the order in which they appear in the archive,
each on a separate line.
If no files are specified, all files in the archive are listed.
.It Fl u
Update files.
When used with the 
.Fl r
option, files in the archive will be replaced
only if the disk file has a newer modification time than the file in
the archive.
When used with the 
.Fl x
option, files in the archive will be extracted
only if the archive file has a newer modification time than the file
on disk.
.It Fl v
Provide verbose output.
When used with the 
.Fl d , 
.Fl m , 
.Fl q 
or 
.Fl x
options,
.Nm ar
gives a file-by-file description of the archive modification.
This description consists of three, white-space separated fields: the
option letter, a dash (``-'') and the file name.
When used with the 
.Fl r
option,
.Nm ar
displays the description as above, but the initial letter is an ``a'' if
the file is added to the archive and an ``r'' if the file replaces a file
already in the archive.
.Pp
When used with the 
.Fl p 
option,
the name of each printed file is written to the standard output before
the contents of the file, preceded by a single newline character, and
followed by two newline characters, enclosed in less-than (``<'') and
greater-than (``>'') characters.
.Pp
When used with the 
.Fl t
option,
.Nm ar
displays an ``ls -l'' style listing of information about the members of
the archive.
This listing consists of eight, white-space separated fields:
the file permissions (see
.Xr strmode 3 ),
the decimal user and group ID's, separated by a single slash (``/''),
the file size (in bytes), the file modification time (in the
.Xr date 1
format ``%b %e %H:%M %Y''), and the name of the file.
.It Fl x
Extract the specified archive members into the files named by the command
line arguments.
If no members are specified, all the members of the archive are extracted into
the current directory.
.Pp
If the file does not exist, it is created; if it does exist, the owner
and group will be unchanged.
The file access and modification times are the time of the extraction
(but see the 
.Fl o
option).
The file permissions will be set to those of the file when it was entered
into the archive; this will fail if the user is not the owner of the
extracted file or the super-user.
.El
.Pp
The
.Nm ar
utility exits 0 on success, and >0 if an error occurs.
.Sh ENVIRONMENT
.Bl -tag -width indent -compact
.It Ev TMPDIR
The pathname of the directory to use when creating temporary files.
.El
.Sh FILES
.Bl -tag -width indent -compact
.It Pa /tmp
default temporary file directory
.It Pa ar.XXXXXX
temporary file names
.El
.Sh COMPATIBILITY
By default,
.Nm ar
writes archives that may be incompatible with historic archives, as
the format used for storing archive members with names longer than
fifteen characters has changed.
This implementation of
.Nm ar
is backward compatible with previous versions of
.Nm ar
in that it can read and write (using the 
.Fl T
option) historic archives.
The 
.Fl T
option is provided for compatibility only, and will be deleted
in a future release.
See
.Xr ar 5
for more information.
.Sh STANDARDS
The
.Nm ar
utility is expected to offer a superset of the 
.St -p1003.2
functionality.
.Sh SEE ALSO
.Xr ld 1 , 
.Xr ranlib 1 ,
.Xr strmode 3 ,
.Xr ar 5


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

Developer/Default.xctoolchain/cctools/ar/ar.c
/*
 * Copyright (c) 1999 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
/*	$OpenBSD: ar.c,v 1.3 1997/01/15 23:42:11 millert Exp $	*/
/*	$NetBSD: ar.c,v 1.5 1995/03/26 03:27:44 glass Exp $	*/

/*-
 * Copyright (c) 1990, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Hugh Smith at The University of Guelph.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
#if 0
static char copyright[] =
"@(#) Copyright (c) 1990, 1993, 1994\n\
	The Regents of the University of California.  All rights reserved.\n";
static char sccsid[] = "@(#)ar.c	8.3 (Berkeley) 4/2/94";
static char rcsid[] = "$OpenBSD: ar.c,v 1.3 1997/01/15 23:42:11 millert Exp $";
#endif
#endif /* not lint */

#include <sys/param.h>

#include <ar.h>
#include <dirent.h>
#include <err.h>
#include <paths.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "archive.h"
#include "extern.h"
#include "stuff/execute.h"
#include "stuff/unix_standard_mode.h"

CHDR chdr;
u_int options;
char *archive, *envtmp, *posarg, *posname;
static void badoptions __P((char *));
static void usage __P((void));
char *progname;

/*
 * main --
 *	main basically uses getopt to parse options and calls the appropriate
 *	functions.  Some hacks that let us be backward compatible with 4.3 ar
 *	option parsing and sanity checking.
 */
int
main(argc, argv)
	int argc;
	char **argv;
{
	int c, retval, verbose, run_ranlib, toc64;
	char *p;
	int (*fcall) __P((char **));

	fcall = 0;
	verbose = 0;
	toc64 = 0;
	progname = argv[0];
	run_ranlib = 1;

	if (argc < 3)
		usage();

	/*
	 * Historic versions didn't require a '-' in front of the options.
	 * Fix it, if necessary.
	*/
	if (*argv[1] != '-') {
		if (!(p = malloc((u_int)(strlen(argv[1]) + 2))))
			err(1, NULL);
		*p = '-';
		(void)strcpy(p + 1, argv[1]);
		argv[1] = p;
	}

	/*
	 * For Rhapsody Premier the option to use long member names, -L, is the
	 * default.  The compiler tools for Rhapsody Premier do understand
	 * extended format #1.  The new option -L allows ar to use the extended 
	 * format and the old -T option causes the truncation of names.
	 */
	while ((c = getopt(argc, argv, "abcdilLmopqrSsTtuVvx6")) != -1) {
		switch(c) {
		case 'a':
			options |= AR_A;
			break;
		case 'b':
		case 'i':
			options |= AR_B;
			break;
		case 'c':
			options |= AR_C;
			break;
		case 'd':
			options |= AR_D;
			fcall = delete;
			break;
		case 'l':		/* not documented, compatibility only */
			envtmp = ".";
			break;
		case 'L':
			options &= ~AR_TR;
			break;
		case 'm':
			options |= AR_M;
			fcall = move;
			break;
		case 'o':
			options |= AR_O;
			break;
		case 'p':
			options |= AR_P;
			fcall = print;
			break;
		case 'q':
			options |= AR_Q;
			fcall = append;
			break;
		case 'r':
			options |= AR_R;
			fcall = replace;
			break;
		case 's':
			options |= AR_S;
			break;
		case 'S':
			options &= ~AR_S;
			run_ranlib = 0;
			break;
		case 'T':
			options |= AR_TR;
			break;
		case 't':
			options |= AR_T;
			fcall = contents;
			break;
		case 'u':
			options |= AR_U;
			break;
		case 'V':
			verbose = 1;
			break;
		case '6':
			toc64 = 1;
			break;
		case 'v':
			options |= AR_V;
			break;
		case 'x':
			options |= AR_X;
			fcall = extract;
			break;
		default:
			usage();
		}
	}

	argv += optind;
	argc -= optind;

	/* One of -dmpqrtsx required. */
	if (!(options & (AR_D|AR_M|AR_P|AR_Q|AR_R|AR_S|AR_T|AR_X))) {
		warnx("one of options -dmpqrtsx is required");
		usage();
	}
	/* Only one of -a and -bi allowed. */
	if (options & AR_A && options & AR_B) {
		warnx("only one of -a and -[bi] options allowed");
		usage();
	}
	/* -ab require a position argument. */
	if (options & (AR_A|AR_B)) {
		if (!(posarg = *argv++)) {
			warnx("no position operand specified");
			usage();
		}
		posname = rname(posarg);
	}
	/* -d only valid with -Tsv. */
	if (options & AR_D && options & ~(AR_D|AR_TR|AR_S|AR_V))
		badoptions("-d");
	/* -m only valid with -abiTsv. */
	if (options & AR_M && options & ~(AR_A|AR_B|AR_M|AR_TR|AR_S|AR_V))
		badoptions("-m");
	/* -p only valid with -Tsv. */
	if (options & AR_P && options & ~(AR_P|AR_TR|AR_S|AR_V))
		badoptions("-p");
	/* -q only valid with -cTsv. */
	if (options & AR_Q && options & ~(AR_C|AR_Q|AR_TR|AR_S|AR_V))
		badoptions("-q");
	/* -r only valid with -abcuTsv. */
	if (options & AR_R && options & ~(AR_A|AR_B|AR_C|AR_R|AR_U|AR_TR|AR_S|AR_V))
		badoptions("-r");
	/* -t only valid with -Tsv. */
	if (options & AR_T && options & ~(AR_T|AR_TR|AR_S|AR_V))
		badoptions("-t");
	/* -x only valid with -ouTsv. */
	if (options & AR_X && options & ~(AR_O|AR_U|AR_TR|AR_S|AR_V|AR_X))
		badoptions("-x");

	if (!(archive = *argv++)) {
		warnx("no archive specified");
		usage();
	}

	/* -dmqr require a list of archive elements. */
	if (options & (AR_D|AR_M|AR_Q|AR_R) && !*argv) {
		warnx("no archive members specified");
		usage();
	}

	if(fcall != 0){
	    retval = (*fcall)(argv);
	    if(retval != EXIT_SUCCESS ||
	       ((options & AR_S) != AR_S &&
	        (get_unix_standard_mode() == FALSE ||
		 archive_opened_for_writing == 0)))
		exit(retval);
	}

	/*
	 * The default is to run ranlib(1) for UNIX conformance.  But if the -S
	 * option is specified by the user we don't run it.
	 */
	if(run_ranlib){
	    /* run ranlib -f or -q on the archive */
	    reset_execute_list();
	    add_execute_list_with_prefix("ranlib");
	    if(options & AR_S)
		add_execute_list("-f");
	    else
		add_execute_list("-q");
	    if(toc64)
		add_execute_list("-toc64");
	    add_execute_list(archive);
	    if(execute_list(verbose) == 0){
		(void)fprintf(stderr, "%s: internal ranlib command failed\n",
			      progname);
		exit(EXIT_FAILURE);
	    }
	}
	exit(EXIT_SUCCESS);
}

static void
badoptions(arg)
	char *arg;
{

	warnx("illegal option combination for %s", arg);
	usage();
}

static void
usage()
{

	(void)fprintf(stderr, "usage:  ar -d [-TLsv] archive file ...\n");
	(void)fprintf(stderr, "\tar -m [-TLsv] archive file ...\n");
	(void)fprintf(stderr, "\tar -m [-abiTLsv] position archive file ...\n");
	(void)fprintf(stderr, "\tar -p [-TLsv] archive [file ...]\n");
	(void)fprintf(stderr, "\tar -q [-cTLsv] archive file ...\n");
	(void)fprintf(stderr, "\tar -r [-cuTLsv] archive file ...\n");
	(void)fprintf(stderr, "\tar -r [-abciuTLsv] position archive file ...\n");
	(void)fprintf(stderr, "\tar -t [-TLsv] archive [file ...]\n");
	(void)fprintf(stderr, "\tar -x [-ouTLsv] archive [file ...]\n");
	exit(1);
}	
